File: luogu/practice/05-string/05-04-Trie/p3879.cpp
#include <iostream>
#include <vector>
using namespace std;

class TrieNode {
public:
    bool isEnd;
    vector<TrieNode*> next;

    TrieNode() {
        isEnd = false;
        next = vector<TrieNode*>(26, nullptr);
    }
};

class Trie {
public:
    TrieNode* root;

    Trie() {
        root = new TrieNode();
    }

    void insert(string word) {
        TrieNode* cur = root;
        for (char c : word) {
            int idx = c - 'a';
            if (cur->next[idx] == nullptr) {
                cur->next[idx] = new TrieNode();
            }
            cur = cur->next[idx];
        }
        cur->isEnd = true;
    }

    bool search(string str, int pos, TrieNode* cur) {
        if (pos == str.length()) {
            return cur->isEnd;
        }
        int idx = str[pos] - 'a';
        if (cur->next[idx] == nullptr) {
            return false;
        }
        return search(str, pos+1, cur->next[idx]);
    }
};

int main() {
    int n;
    cin >> n;
    vector<Trie*> tries(n);
    for (int i = 0; i < n; i++) {
        tries[i] = new Trie();
        int k;
        cin >> k;
        for (int j = 0; j < k; j++) {
            string s;
            cin >> s;
            tries[i]->insert(s);
        }
    }
    int m;
    cin >> m;
    for (int i = 0; i < m; i++) {
        string x;
        cin >> x;
        for (int j = 0; j < n; j++) {
            if (tries[j]->search(x, 0, tries[j]->root)) {
                cout << j+1 << " ";
            }
        }
        cout << endl;
    }
    return 0;
}

File: luogu/practice/05-string/05-04-Trie/p3879.py
class TrieNode:
    def __init__(self):
        self.isEnd = False
        self.next = [None] * 26

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        cur = self.root
        for c in word:
            idx = ord(c) - ord('a')
            if cur.next[idx] is None:
                cur.next[idx] = TrieNode()
            cur = cur.next[idx]
        cur.isEnd = True

    def search(self, str, pos, cur):
        if pos == len(str):
            return cur.isEnd
        
        idx = ord(str[pos]) - ord('a')
        if cur.next[idx] is None:
            return False

        return self.search(str, pos+1, cur.next[idx])


n = int(input())
tries = [Trie() for _ in range(n)]

for i in range(n):
    strs = input().split()[1:]
    for s in strs:
        tries[i].insert(s)

m = int(input())
for i in range(m):
    x = input()
    for j in range(n):
        if tries[j].search(x, 0, tries[j].root):
            print(j + 1, end=" ")
    print("")

